radix-engine/src/state_manager/staging.rs

```diff
@@ -1,189 +1,581 @@
+                }
 
-impl<'t, 's, S: ReadableSubstateStore> StagedSubstateStore<'t, 's, S> {
-    fn get_substate_recurse(&self, substate_id: &SubstateId, id: u64) -> Option<OutputValue> {
-        if id == 0 {
-            return self.stores.root.get_substate(substate_id);
-        }
+                let new_root = self.nodes.get(new_root_key).unwrap();
+                // Reparent to new_root node and delete it
+                self.children_keys = new_root.children_keys.clone();
+                for key in self.children_keys.iter() {
+                    let node = self.nodes.get_mut(*key).unwrap();
+                    node.parent_key = StagedSubstateStoreKey::RootStoreKey;
+                }
 
-        let node = self.stores.nodes.get(&id).unwrap();
-        if let Some(output) = node.outputs.get(substate_id) {
-            return Some(output.clone());
-        }
+                Self::remove_node(
+                    &mut self.nodes,
+                    &mut self.total_weight,
+                    visitor,
+                    &new_root_key,
+                );
 
-        self.get_substate_recurse(substate_id, node.parent_id)
+                // If the number of state changes that overlap with the self.root (dead_weight) store is greater
+                // than the number of state changes applied on top of it (total_weight), we recalculate the
+                // ImmutableStores in order to free up memory.
+                if self.dead_weight > self.total_weight {
+                    self.recompute_data();
+                }
+            }
+        }
     }
 }
 
-impl<'t, 's, S: ReadableSubstateStore> ReadableSubstateStore for StagedSubstateStore<'t, 's, S> {
-    fn get_substate(&self, substate_id: &SubstateId) -> Option<OutputValue> {
-        self.get_substate_recurse(substate_id, self.id)
-    }
+pub struct StagedSubstateStore<'t, S: ReadableSubstateStore> {
+    manager: &'t StagedSubstateStoreManager<S>,
+    key: StagedSubstateStoreKey,
 }
 
-impl<'t, 's, S: ReadableSubstateStore> WriteableSubstateStore for StagedSubstateStore<'t, 's, S> {
-    fn put_substate(&mut self, substate_id: SubstateId, output: OutputValue) {
-        if self.id == 0 {
-            panic!("Cannot write to root store");
-        } else {
-            let node = self.stores.nodes.get_mut(&self.id).unwrap();
-            node.outputs.insert(substate_id, output);
+impl<'t, S: ReadableSubstateStore> ReadableSubstateStore for StagedSubstateStore<'t, S> {
+    fn get_substate(&self, substate_id: &SubstateId) -> Option<OutputValue> {
+        match self.key {
+            StagedSubstateStoreKey::RootStoreKey => self.manager.root.get_substate(substate_id),
+            StagedSubstateStoreKey::InternalNodeStoreKey(key) => {
+                match self
+                    .manager
+                    .nodes
+                    .get(key)
+                    .unwrap()
+                    .store
+                    .get_substate(substate_id)
```

Comment:
It seems that we do not merge the updates from the receipt into the store in `new_child_node`. Are we sure  this will return the correct substate? Suggest unit testing.

Reply:
We do in `StagedSubstateStoreNode::new` called by `StagedSubstateStoreManager::new_child_node`. Trying to write unit test.

Reply:
Sounds good!

Reply:
I've updated the unit test so that it's easy to modify in terms of the structure that it's building and also to check for ALL staged stores that have been built every time a new one is added.